arch/arm/core/userspace.S

```diff
@@ -0,0 +1,178 @@
+/*
+ * Userspace and service handler hooks
+ *
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+#include <offsets_short.h>
+#include <toolchain.h>
+#include <linker/sections.h>
+#include <kernel_structs.h>
+#include <arch/cpu.h>
+#include <syscall.h>
+
+_ASM_FILE_PROLOGUE
+
+GTEXT(_arm_userspace_enter)
+GTEXT(_arm_do_syscall)
+GDATA(_kernel)
+
+/* Imports */
+GTEXT(_k_syscall_table)
+
+/**
+ *
+ * User space entry function
+ *
+ * This function is the entry point to user mode from privileged execution.
+ * The conversion is one way, and threads which transition to user mode do
+ * not transition back later, unless they are doing system calls.
+ *
+ */
+SECTION_FUNC(TEXT,_arm_userspace_enter)
+    /* preserve arguments */
+    mov lr, r0
+
+    /*
+     * calculate end of stack and set SP to it
+     */
+    ldr r0, [sp]
+    ldr ip, [sp,#4]
+    add r0, ip
+
+    msr PSP, r0
+
+    /* push args to stack */
+    push {r1,r2,r3,lr}
+
+    /* clear the stack area to clean out privileged data */
+    /* from right past the guard right up to the args */
+    mrs r0, PSP
+#ifdef CONFIG_INIT_STACKS
+    ldr r1,=0xaaaaaaaa
+#else
+    ldr r1,=0
+#endif
+    mov r2, ip
+    sub r2, #16		/* don't erase the pushed regs */
+#ifdef CONFIG_MPU_STACK_GUARD
+    ldr r3, =MPU_GUARD_ALIGN_AND_SIZE
+    sub r2, r3
+    add r0, r3
+#endif
+    bl memset
+
+    /* setup arguments to thread_entry function */
+    ldr r0, =_kernel
+    ldr r0, [r0, #_kernel_offset_to_current]
+    bl configure_mpu_user_stack_context
+
+    pop {r0,r1,r2,lr}
+    /* change processor mode to unprivileged */
+    mov ip, #3
+    msr CONTROL, ip
+
+    /* jump to user entry */
```

Comment:
I might be mis-reading this, but it appears we are directly diving into the supplied user_entry function.
We shouldn't be doing that.
We should be "returning" to the kernel function _thread_entry() define in kernel/thread.c with user_entry, arg0, arg1, arg2 as its arguments.

Reply:
How would we return to "_thread_entry()"?  We never invoked that function.  We set the entry to user_mode_enter, which calls the asm function which in turn calls the user supplied function.  The arch_user_mode_enter() is basically a copy of the _thread_entry, except it calls the asm function to drop to user mode and jump to the supplied function.  Are you implying that we need to call _thread_entry after dropping to user mode?  Which would mean returning back to the arch user mode enter, and then calling _thread_entry?